src/rate_limiter/src/lib.rs

```diff
@@ -357,23 +385,38 @@ impl RateLimiter {
             TokenType::Ops => self.ops.as_mut(),
         };
         // Try to consume from the token bucket.
-        let success = match token_bucket {
-            Some(bucket) => bucket.reduce(tokens),
+        match token_bucket {
+            Some(bucket) => {
+                let refill_time = bucket.refill_time_ms();
+                match bucket.reduce(tokens) {
+                    // When we report budget is over, there will be no further calls here,
+                    // register a timer to replenish the bucket and resume processing;
+                    // make sure there is only one running timer for this limiter.
+                    BucketReduction::Failure => {
+                        self.activate_timer(None);
```

Comment:
Ties into the comment above: clearer to do:
```suggestion
                        self.activate_timer(TIMER_REFILL_STATE);
```

Reply:
Done